src/nnn.c

```diff
@@ -8001,28 +8001,26 @@ static char *load_input(int fd, const char *path)
 			off = next - input;
 		}
 
-		if (chunk_count == 512) {
-			msgnum = MSG_SIZE_LIMIT;
-			break;
-		}
-
 		/* We don't need to allocate another chunk */
-		if (chunk_count == (total_read - input_read) / chunk)
+		if (chunk_count > (total_read + chunk - 1) / chunk)
 			continue;
 
-		chunk_count = total_read / chunk;
-		if (total_read % chunk)
-			++chunk_count;
+		/* We can never need more than one additional chunk */
+		++chunk_count;
+		if (chunk_count > 512) {
+			msgnum = MSG_SIZE_LIMIT;
+			break;
+		}
 
-		input = xrealloc(input, (chunk_count + 1) * chunk);
+		input = xrealloc(input, chunk_count * chunk);
 		if (!input)
-			return NULL;
+			goto malloc_1;
 	}
 
 	/* Read off the extra data if limits exceeded */
 	if (msgnum) {
 		char buf[512];
-		while (read(fd, buf, 512) > 0);
+		while (read(fd, buf, 512) != 0);
```

Comment:
Would prefer using `sizeof` instead of hard coding.

```suggestion
		while (read(fd, buf, sizeof buf) != 0);
```

There's a lot of other places in our code that uses hard-coded sizes where sizeof would've sufficed (and would've been more robust against future changes).

Reply:
Now that I'm looking at it again, not sure if this is correct either.. This *might* end up going into an infinite loop where read fails for some reason and keeps returning `-1`.

Reply:
I suppose something like this would be more robust and deal with edge case of signal interrupt as well.

```c
ssize_t n;
do {
	if ((n = read(...)) < 0 && errno != EINTR)
		break;
} while (n);
```

Reply:
Is that what we are going with?

Reply:
I mean this is a dumb question, but what do we do if we are given an endless input? Like `/dev/zero` or something simply huge. That shouldn't ever happen even on accident, but do we really just try to read it endlessly until the user aborts?

Reply:
> Is that what we are going with?

Not sure. The upper `read` doesn't deal with `EINTR` either. So I guess we can leave it at `> 0`.

> I mean this is a dumb question, but what do we do if we are given an endless input?

Any specific reason we're draining the input to begin with? What happens if we don't drain it?

Reply:
Also couldn't we just close the fd? We reopen `stdin` after the call to `load_input()` either way. Closing the fd would offload the task of dealing with the excess data to the kernel. And it can surely handle it much more efficiently.

Reply:
> Also couldn't we just close the fd? We reopen stdin after the call to load_input() either way

I'd go with that approach as well instead of trying to drain the input.

Reply:
We are doing that. Kind of a weird decision to leave closing of the fd to the caller of the function, but it's called from `readpipe()` as well, which seems to also delegate the responsibility to its caller who finally is the one that opens and closes it. I'm not sure about all that. The other path through main is much simpler.

Also this just looks off to me. How is this even valid syntax.
```c
do
	rfd = open(g_pipepath, O_RDONLY);
while (rfd == -1 && errno == EINTR);
```

Reply:
> Also this just looks wrong to me. How is this even valid syntax.

It's just omitting the  `{}` cus it's a single statement. I've seen (and done) worse stuff...

```c
	n[(char *)memcpy(d, s, n)] = '\0';
```

Reply:
> It's just omitting the {} cus it's a single statement.

Oh, I know why it is. But I don't really think it should.

I do like the `3[a]` way of indexing. And if it weren't hard to read that particular statement looks pretty nice. The worst I've done is something along the lines of
```c
ss = strstr(s, "...");
if (ss && ss != s && ss[-1] != '/') ...
```